src/Avalonia.Controls/TextBox.cs

```diff
@@ -1063,6 +1063,14 @@ private void HandleTextInput(string? input)
                     _undoRedoHelper.DiscardRedo();
                 }
 
+                //Make sure updated text is in sync
+                _presenter?.SetCurrentValue(TextPresenter.TextProperty, text);
+
+                caretIndex += input.Length;
```

Comment:
`input.Length` is added twice, here and on `SetCurrentValue(CaretIndexProperty..., )` below, moving the effective caret twice (the visual one is correct).

Shouldn't we add some small unit test ensuring that `TextBox.CaretIndex == _presenter.CaretIndex` in the bugged cases after input?

Reply:
Good catch. These unit tests tend to be hard to write because of the property sync. I am having a look. All these issues are caused by the fact that some state is synced via bindings.